Feature rhino turn with encoder

crisvdn opened the pull request:
This pull request implements a working rhino class. This rhino class can turn with the encoders and drive forward/ backwards.

crisvdn committed:
Missing encoders in constructor of beetle

crisvdn committed:
Integrated existing code from feature-rhino-turn with most up-to-date release.

crisvdn committed:
Added encode values in constructor of beetle.

crisvdn committed:
added test main loop

crisvdn committed:
Inverted speed of 1 motor so the rhino drives forward/backward properly.

crisvdn committed:
Turn function nowuses encoders to turn.

Joepieler committed:
Merge pull request #17 from R2D2-2019/release

merge Release in to Master

crisvdn committed:
Implemented turn function with encoders

The rhino class is now compatible with the new updated UML class diagram. The rhino now uses the encoders to turn in a circle.

crisvdn committed:
Edited main test loop

Test loop for rhino turn is inside this main. Added documentation for each line what is done.

crisvdn committed:
Merge branch 'release' into feature-rhino-turn-with-encoder

crisvdn committed:
Refactored rhino class to be compatible after merge

Small changes were done to interface class due to a change in class diagram. After this merge, functions were edited / removed which were removed in the rhino class in this commit.

crisvdn committed:
Merge branch 'master' into feature-rhino-turn-with-encoder

Joepieler commented on code/main.cpp:
```diff
@@ -1,35 +1,54 @@
     r2d2::comm_c comm;
 
-    // Create The moving platform the Beetle
-    auto beetle =
```
a qik can't be a beetle and a rhino at the same time

crisvdn replied:
Fixed.

Joepieler commented on code/main.cpp:
```diff
@@ -1,35 +1,54 @@
-    while (true) {
-        beetle.process();
-        hwlib::wait_ms(100);
+    if (rhino_motor_test) {
```
No tests in the main of the release.

crisvdn replied:
Fixed.

Joepieler commented on code/src/rhino.cpp:
```diff
@@ -0,0 +1,97 @@
+        bool low_m1 = false;
+        int counter_m1 = 0;
+        // The adc input. is between 3000 and 3800.
+        unsigned int adc_voltage = 3500;
```
adc_voltage can be constexpr

crisvdn replied:
This variable is now constexpr static.

Joepieler commented on code/src/rhino.cpp:
```diff
@@ -0,0 +1,97 @@
+            }
+
+            if (counter_m0 ==
+                (int(encode_1_full_turn * turn_factor / 360 * degrees))) {
```
can be calculated before the loop

crisvdn replied:
This calculation is now stored in a variable.

Joepieler commented on code/src/rhino.cpp:
```diff
@@ -0,0 +1,97 @@
+        qik_2s12v10_motorcontroller.set_m0_speed(turn_speed);
+        qik_2s12v10_motorcontroller.set_m1_speed(-1 * turn_speed);
+
+        while (true && degrees != 0) {
```
the True can go away.

crisvdn replied:
This has been changed.

crisvdn committed:
Changed function variables to private varables

min_degrees and max_degrees are now constexpr and uint8_t to maximize the amount of memory usage

crisvdn committed:
adc_voltage is now constexpr

Changes made that were requested by PR

crisvdn committed:
Removed meaningless comments

crisvdn committed:
removed test code in main

The code is now ready and only uses the process function. Unused boolean also removed

crisvdn committed:
Merge branch 'feature-rhino-turn-with-encoder' of https://github.com/R2D2-2019/moving_platform into feature-rhino-turn-with-encoder

crisvdn committed:
Refactored calculation into variable

Calculation was done multiple times has been replaced by encoder_target_value. This variable holds the amount of revolutions the encoder needs to reach to make sure the turn of the rhino is complete

crisvdn committed:
Constexpr values are now static

A compile error was generated, this commit fixes the compile error about constexpr static variables.

crisvdn committed:
Removed unnecessary condition in while loop

crisvdn committed:
Added documentation inside turn function

Added documentation on the booleans inside the turn function.

crisvdn committed:
Changed while condition

Changed while condition so the code is more efficient and has one if statement less to check.

Joepieler reviewed (approved):
Everything i asked for has been changed.
